### PR TITLE
Refactor/lock and runner module

### PR DESCRIPTION
**Pull request by jackey8616**

### CODE REVIEW

File: src/main.ts [hunk 1]
```diff
@@ -17,24 +18,26 @@ async function bootstrap() {
...
 
-    app.useGlobalFilters(new ApiExceptionFilter());  
-    port = 8081;
+    app = app.useGlobalFilters(new ApiExceptionFilter());  
   }
   
-  app.enableShutdownHooks();
-  await app.listen(port);
+  app = app.enableShutdownHooks();
+  app = await app.listen(8081);
```
Comment (kkshyu) [inline src/main.ts]: it is better to move port into env

File: src/runner/runner.module.ts [hunk 2]
```diff
@@ -1,29 +1,28 @@
...
 import { Runner } from './runner';
 import { RunnerService } from './runner.service';
-import { RunnerType } from './runner.type';
+import { RunnerController } from './runner.controller';
 
 @Module({})
 export class RunnerModule {
   static forRoot(options: {
-    workerName: string; nodeEnv: string;
+    workerName: string; nodeEnv: string; clazz: any;
```
Comment (kkshyu) [inline src/runner/runner.module.ts]: change into Runner type

File: src/runner/runner.service.ts [hunk 3]
```diff
@@ -1,15 +1,29 @@
...
+    const now = new Date();
+    const previousExecutedTime = this.runner.getPreviousExecutedTime();
+    const runnerInterval = this.runner.getInterval();
+
+    if (!previousExecutedTime) {
+      return 'not execute yet';
+    } else if (now.getTime() - previousExecutedTime.getTime() > runnerInterval) {
+      throw new APIException({ code: 'E_HEALTHZ', message: 'Runner is hang...' }, 500);
+    }
+    return new Date().toISOString();
```
Comment (kkshyu) [inline src/runner/runner.service.ts]: change into `previousExecutedTime` to show more info for debugging

File: src/runner/runner.ts [hunk 4]
```diff
@@ -1,53 +1,60 @@
...
-    if (isAcquiredLock === false) {
-      console.error(`Unable to acquire lock for ${this.name}:${this.uuid}`);
-    } else if (this.isCompleted === false) {
-      console.error('Previous execution is not completed yet');
-      return;
-    }
-    
     try {
-      this.isCompleted = false;
+      await this.lockService.acquireLock(
```
Comment (kkshyu) [inline src/runner/runner.ts]: please use `occupyLock` for better naming

File: src/utility/lock/distributed_lock.service.ts [hunk 5]
```diff
@@ -6,62 +6,39 @@ import { CacheService } from '../cache/cache.service';
...
+    options?: { subKey: string },
+  ): Promise<void> {
+    const redisCli = this.cacheService.getClient();
+    const key = `lock:${this.key}${options ? `:${options.subKey}` : ''}`;
+    const inRedisKeys = await redisCli.keys(`${key}:*`);
+    
+    if (inRedisKeys.length >= this.maxHolderAmount) {
+      throw new Error(
+        `Lock [${key}] with identity [${identityKey}] is unable to acquired.`
+      );
```
Comment (kkshyu) [inline src/utility/lock/distributed_lock.service.ts]: remove checking keys existence. only accept EXACTLY one process to access the resource.

File: src/utility/lock/distributed_lock.service.ts [hunk 6]
```diff
@@ -6,62 +6,39 @@ import { CacheService } from '../cache/cache.service';
...
     }
+    await redisCli.set(`${key}:${identityKey}`, value, 'PX', expireTime, 'NX');
```
Comment (kkshyu) [inline src/utility/lock/distributed_lock.service.ts]: if we cannot set the lock, throw error for occupying failed

File: src/utility/lock/distributed_lock.service.ts [hunk 7]
```diff
@@ -6,62 +6,39 @@ import { CacheService } from '../cache/cache.service';
...
-    } catch {
-      return false;
-    }
-  }
+  async releaseLock(
+    identityKey: string, options?: { subKey: string; },
+  ): Promise<void> {
+    const key = `lock:${this.key}${options ? `:${options.subKey}` : ''}:${identityKey}`;
+    const redisCli = this.cacheService.getClient();
+    const inRedisValue = await redisCli.exists(key);
```
Comment (kkshyu) [inline src/utility/lock/distributed_lock.service.ts]: do not need to check the existence. just delete the lock key in any case. if we do the existence check, may cause extra issue(like operation failed...)

File: src/utility/lock/lock.module.ts [hunk 8]
```diff
@@ -8,17 +8,16 @@ import { DistributedLockService } from './distributed_lock.service';
 export class LockModule {
   static forFeature(options: {
     key: string;
+    maxHolderAmount: number;
```
Comment (kkshyu) [inline src/utility/lock/lock.module.ts]: we don't need maxHolderAmount for this case

File: test/runner.e2e-spec.ts [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+        .enableShutdownHooks();
+      const app2 =moduleFixture2
+        .createNestApplication()
+        .enableShutdownHooks();
+
+      app1.get(ShutdownService).subscribeToShutdown(async () => await app1.close());
+      app2.get(ShutdownService).subscribeToShutdown(async () => await app2.close());
+      await app1.init();
+      // Interleaved initialization
+      await new Promise((resolve) => setTimeout(() => resolve(true), 500));
```
Comment (kkshyu) [inline test/runner.e2e-spec.ts:126]: set cases from 0, 100, 500, 1000

File: src/runner/runner.ts [hunk 10]
```diff
@@ -1,53 +1,60 @@
...
+        this.uuid,
+        new Date().getTime(),
+        this.interval * 2,
+      );
+      const suicideTimeout = setTimeout(
+        async () => await this.suicide(), this.interval * 1.1,
+      );
```
Comment (kkshyu) [inline src/runner/runner.ts]: move to the top -> if process cannot occupy any lock, then go to suicide

**Review by kkshyu — CHANGES_REQUESTED** [review]

File: src/main.ts (hunk 1 above)
Comment (jackey8616) [reply src/main.ts]: Fixed at: [e4ea457](https://github.com/urfit-tech/lodestar-server/commit/e4ea457)

**Review by kkshyu — APPROVED** [review]

**Pull request merged**